topydo/ui/TodoListWidget.py

```diff
@@ -101,132 +119,199 @@ def _scroll_to_bottom(self, p_size):
-            if p_key not in ['d', 'w', 'm', 'y']:
-                if p_key.isdigit():
-                    self._pp_offset += p_key
-                elif self.keystate == 'p' and p_key == 's':
-                    self.keystate = 'ps'
-                else:
-                    self._pp_offset = ''
-                    self.keystate = None
+        keymap, keystates = self.keymap
+
+        shortcut = self.keystate or ''
+        if len(p_key) > 1:
+            shortcut += translate_key_to_cfg(p_key)
+        else:
+            shortcut += p_key
+
+        try:
+            action = keymap[shortcut]
+        except KeyError:
+            action = None
+
+        if action:
+            if shortcut in keystates:
+                # Supplied key-shortcut matches keystate and action. Save the
+                # keystate in case user will hit another key and add an action
+                # waiting for execution if user won't type anything further.
+                self.keystate = shortcut
+                if action not in requires_further_input:
+                    self._add_pending_action(action, p_size)
             else:
-                self._postpone_selected_item(p_key)
-                self._pp_offset = ''
+                # Only action is matched. Handle it and reset keystate.
+                self.action_handler(action, p_size)
                 self.keystate = None
-
-            return
-        elif self.keystate == 'r':
-            if p_key.isalpha():
-                self._pri_selected_item(p_key)
-            self.keystate = None
             return
-
-        if p_key == 'x':
-            self._complete_selected_item()
-        elif p_key == 'p':
-            self.keystate = 'p'
-        elif p_key == 'd':
-            self._remove_selected_item()
-        elif p_key == 'e':
-            self._edit_selected_item()
-            # force screen redraw after editing
-            urwid.emit_signal(self, 'refresh')
-        elif p_key == 'r':
-            self.keystate = 'r'
-        elif p_key == 'u':
-            urwid.emit_signal(self, 'execute_command', "revert")
-        elif p_key == 'j':
-            self.listbox.keypress(p_size, 'down')
-        elif p_key == 'k':
-            self.listbox.keypress(p_size, 'up')
-        elif p_key == 'home':
-            self._scroll_to_top(p_size)
-        elif p_key == 'G' or p_key == 'end':
-            self._scroll_to_bottom(p_size)
-        elif p_key == 'g':
-            self.keystate = 'g'
         else:
-            return self.listbox.keypress(p_size, p_key)
+            if shortcut in keystates:
+                self.keystate = shortcut
+            else:
+                try:
+                    # Check whether current keystate matches built-in 'postpone'
+                    # action.
+                    pp_mode = keymap[self.keystate]
+                    if pp_mode in ['postpone', 'postpone_s']:
+                        if self._postpone_selected(p_key, pp_mode) is not None:
```

Comment:
After [recent change](https://github.com/bram85/topydo/pull/101/files#diff-b9959c14ca7b8926ee9d0fee31de25d4R160) I would leave it as is.